embrace-android-sdk/src/main/cpp/file_writer.c

```diff
@@ -158,147 +165,313 @@ emb_error *emb_read_errors_from_file(const char *path) {
+        goto Exit;
+    }
+    crash_object = json_value_get_object(crash_value);
+    if (crash_object == NULL) {
+        goto Exit;
+    }
 
-    EMB_LOGDEV("Converting tree to JSON string.");
-    char *serialized_crash = json_serialize_to_string_pretty(crash_value);
+    // start adding metadata/basic info to tree
+    if (!emb_add_metadata_to_json(crash, root_object)) {
+        goto Exit;
+    }
+    if (!emb_add_basic_info_to_json(crash, root_object)) {
+        goto Exit;
+    }
 
-    EMB_LOGDEV("Starting Base64 encoding.");
-    char *base64_crash = b64_encode(serialized_crash, strlen(serialized_crash));
-    json_free_serialized_string(serialized_crash);
+    // add crash data to tree
+    if (!emb_build_crash_json_tree(crash, root_object, crash_object)) {
+        goto Exit;
+    }
 
-    EMB_LOGDEV("Altering JSON tree root.");
-    json_object_set_string(root_object, kCrashKey, base64_crash);
-    free(base64_crash);
+    if (!emb_add_b64_value_to_json(root_object, crash_value)) {
+        goto Exit;
+    }
 
-    // final result
+    // serialize final result & return as string
     EMB_LOGDEV("Serializing final JSON string");
-    serialized_string = json_serialize_to_string_pretty(root_value);
-//    json_free_serialized_string(serialized_string);
-    json_value_free(root_value);
-    json_value_free(crash_value);
+    serialized_string = json_serialize_to_string(root_value);
+
+    Exit:
+    if (root_value != NULL) {
+        json_value_free(root_value);
+    }
+    if (crash_value != NULL) {
+        json_value_free(crash_value);
+    }
     return serialized_string;
 }
 
-void emb_add_metadata_to_json(const emb_crash *crash, JSON_Object *root_object) {
+bool emb_add_metadata_to_json(const emb_crash *crash, JSON_Object *root_object) {
     JSON_Value *meta_value = json_parse_string(crash->meta_data);
+    bool success = false;
 
     if (meta_value != NULL) {
-        EMB_LOGDEV("Successfully parsed crash JSON metadata");
-        json_object_set_value(root_object, kDeviceMetaKey, meta_value);
-    } else {
-        EMB_LOGERROR("Could not JSON decode metadata: %s", crash->meta_data);
+        success = json_object_set_value(root_object, kDeviceMetaKey, meta_value) == JSONSuccess;
     }
+    return success;
 }
 
-void emb_add_basic_info_to_json(const emb_crash *crash, JSON_Object *root_object) {
+bool emb_add_basic_info_to_json(const emb_crash *crash, JSON_Object *root_object) {
     EMB_LOGDEV("Serializing IDs + payload version.");
-    json_object_set_string(root_object, kReportIDKey, crash->report_id);
-    json_object_set_string(root_object, kVersionKey, kCurrentPayloadVersion);
-    json_object_set_number(root_object, kCrashTSKey, crash->crash_ts);
-    json_object_set_string(root_object, kSessionIDKey, crash->session_id);
-    json_object_set_string(root_object, kAppStateKey, crash->app_state);
+    if (json_object_set_string(root_object, kReportIDKey, crash->report_id) != JSONSuccess) {
```

Comment:
Could we create a function that compares with JSONSuccess and returns true/false

```
bool set_string(object, key, value) {
    if (json_object_set_number(object, key, value) != JSONSuccess) {
        return false;
    }
    return true;
}
```
And then do something like
```
if (!set_string(root_object, kReportIDKey, crash-report_id) || !set_string(root_object, kVersionKey, kCurrentPayloadVersion) || etc...) 
    return false 
```

So we avoid all the `return false;`

Reply:
or add something like this to wrap functions

```
bool isJSONSuccess(JSON_Status result) {
    return result == JSONSuccess ? true : false;
}
```

Reply:
I've updated the PR to use a macro

Reply:
> I've updated the PR to use a macro

looks good!